### PR TITLE
allow non-signed in users to do channel backed puzzles

### PR DESCRIPTION
**Pull request by Bjvanminnen**
No longer require users to be logged in to do channel-backed levels. This is the payoff PR from a lot of previous work, most recently https://github.com/code-dot-org/code-dot-org/pull/18329.

Once this is in place, it enables a few other pieces of work
1. https://github.com/code-dot-org/code-dot-org/pull/18299 An LB setting that allows us to mark scripts as being age 13+ required, but not login required. This will be used on the Applab HOC script.
2. We can delete a couple of columns from channel_tokens, namely user_id and channel. I'll probably hold off a couple days on this just to make sure nothing has gone south.

### CODE REVIEW

File: dashboard/app/helpers/levels_helper.rb [hunk 1]
```diff
@@ -46,10 +46,6 @@ def readonly_view_options
   # Returns the channel associated with the given Level and User pair, or
   # creates a new channel for the pair if one doesn't exist.
   def get_channel_for(level, user = nil)
```
Comment (davidsbailey) [inline dashboard/app/helpers/levels_helper.rb]: The function level comment says we're creating a new channel if one doesn't exist, but in the case that user is not null, it looks like we are only searching for an existing token (in `ChannelToken.find_channel_token`), not creating a new one. It seems like we'd need to create a new one if one did not exist. Am I missing something?

Comment (Bjvanminnen) [reply dashboard/app/helpers/levels_helper.rb]: I think some of the confusion comes from differentiating between user and current_user. I'll try and make the comment better.

In the case where we're passed a user (not current_user), it means we have a teacher looking at a student's level. In this case we will never create a new channel.

In the case that we have a current_user, we will look for an existing channel_token and create one if it does not exist.

In the case that current_user is nil, the same thing will happen, except that `storage_id_for_user_id` will potentially end up generating a new storage_id for this anon user.

**Comment by davidsbailey** [comment]
OK, this makes sense. It could be worth a quick manual check that a new, signed-out user is receiving the newly-generated storage id, and that the same storage id is being sent back to the server after navigating away and back.

LGTM after considering the `safe_get_channel_for` issue, which could be addressed in a different PR.

**Comment by Bjvanminnen** [comment]
`safe_get_channel_for` was removed in this PR https://github.com/code-dot-org/code-dot-org/commit/1af210af702ca943e1f2b09ac8221e6873015f0a#diff-1f68e16ab551449bdbd0fb7f01004f68, but apparently I somehow missed a usage. Concerning that this wasn't noticed before now. Nice catch. I will open a new PR to address that.

**Pull request merged**